Cmake yaml module

tejlmand opened the pull request:
This PR introduces a CMake yaml module to facilitate reading and writing of simple yaml files directly in CMake.

Zephyr uses yaml intensively, and in CMake there is already several places where basic yaml files are written.

Having a common module helps us to cleanup and align yaml usage in Zephyr CMake and make future use-cases easier to implement.

tejlmand committed:
cmake: initial version of a CMake yaml module

Initial CMake yaml module to facilitate reading yaml files into CMake,
update yaml settings and write it back to a file.

The yaml module also supports creation of yaml files from scratch.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
tests: cmake: add test cases for CMake YAML module

This commit provides a basic set of tests for Zephyr's CMake YAML
module.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nordicjm commented on cmake/modules/yaml.cmake:
```diff
@@ -0,0 +1,403 @@
+  zephyr_set(FILE ${ARG_YAML_FILE} SCOPE ${ARG_YAML_NAME})
+
+  execute_process(COMMAND python -c
+    "import json; import yaml; print(json.dumps(yaml.safe_load(open('${ARG_YAML_FILE}'))))"
```
(not requesting change) something to probably thing about is that this escapes from of cmake into python as another process, from a security perspective anyone can put any file that is visible to the OS and might allow some level of privileged file access

tejlmand replied:
The `ARG_YAML_FILE` is passed as an argument to `yaml_load()`, so true that a caller of said function can do:
`yaml_load(NAME foo FILE "<insert python escape command>")`.

But if the caller of `yaml_load()` can do that, then that caller can much easier call `execute_process()` directly without the hassle of going through another function. Just call: `execute_process(COMMAND python -c "...")` directly.

So I guess you're basically saying that CMake is unsafe, especially if you use `sudo cmake` :wink:

nordicjm commented on cmake/modules/yaml.cmake:
```diff
@@ -0,0 +1,403 @@
+#   yaml_write(NAME <name>)
+#
+# Write the YAML context <name> to the file which were given with the earlier
+# yaml_load() or yaml_create() call.
```
```suggestion
# 'yaml_load()' or 'yaml_create()' call.
```

pdgendt commented on cmake/modules/yaml.cmake:
```diff
@@ -0,0 +1,403 @@
+#
+# NAME <name>: Name of the YAML context
+#
+function(yaml_write)
```
Wouldn't it make sense to decouple the context from the file? And add a `yaml_save` function that can be passed a filename?

tejlmand replied:
I guess it depends on the use-case.
The two use-cases I currently had in mind were
1. you know everything at one location, like the [domains.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/share/sysbuild/cmake/domains.cmake) file.
    There it doesn't really matter.
2. You may know about the file (and load it into a context) at one location, but you'll be updating the context at other locations, then all you want to know is the agreed context name so that you can. You don't want/need to know about file names. That's why the actual filename is a property on the context.

but I see your proposal / use-case as just as valid, so perhaps allowing for either write to specific filename, or omitting the filename when creating a new context.

In short:
- allow specifying a file name on `yaml_write()`
  The filename property of the context will be used when filename is not provided with `yaml_write`.
    - Fail with error if calling `yaml_write()` if context has no filename property defined and filename was not passed as argument.
- make filename optional on `yaml_create()`

pdgendt replied:
I think these additions are welcome, can we rename `yaml_write` to `yaml_save`? If you use `load`, then `save` seems better suited.

Would it also make sense to allow loading multiple files to amend/overwrite values in the current context? For this the decoupling makes over more sense.

tejlmand replied:
> can we rename yaml_write to yaml_save? If you use load, then save seems better suited.

Good point and I agree, will do.

> Would it also make sense to allow loading multiple files to amend/overwrite values in the current context? For this the decoupling makes over more sense.

I would leave that for a future enhancement.
But one use-case (not yet supported) which I consider useful, is to be able to create a new context and then merge that into a context loaded from file. That principle is very similar to the loading of multiple files.

tejlmand replied:
> I think these additions are welcome, can we rename yaml_write to yaml_save?

Done

pdgendt commented on cmake/modules/yaml.cmake:
```diff
@@ -0,0 +1,414 @@
+endfunction()
+
+# Usage:
+#   yaml_create(FILE <file> NAME <name>)
```
`File <file>` is optional now.

tejlmand replied:
Fixed

pdgendt commented on cmake/modules/yaml.cmake:
```diff
@@ -0,0 +1,414 @@
+  zephyr_set(JSON "${json_content}" SCOPE ${ARG_YAML_NAME})
+endfunction()
+
+# Usage:
+#   yaml_save(NAME <name>)
+#
+# Write the YAML context <name> to the file which were given with the earlier
+# 'yaml_load()' or 'yaml_create()' call.
+#
+# NAME <name>: Name of the YAML context
+#
```
Missing `File` argument info.

tejlmand replied:
Fixed

tejlmand committed:
cmake: add check_arguments_... macro to facilitate function writing

Add a new zephyr_check_arguments_required_allow_empty() macro for
function argument validation.

Zephyr already has a zephyr_check_arguments_required() for checking
required arguments, like
zephyr_check_arguments_required(foo_func <prefix> FOO BAR)
which ensures that one of FOO or BAR are given in a call like:
foo_func(BAR val)

One limitation however, is that is some cases BAR may be allowed to be
empty, so that it's still possible to know if FOO or BAR were supplied.
In most case, BAR and FOO will have values following the keyword, like:
foo_func(BAR my_bar_val)
foo_func(FOO my_foo_val)

but in cases where `my_bar_val` is a variable which may be empty, like:
set(my_bar_val)
foo_func(BAR ${my_bar_val}) # (expands to: foo_func(BAR)

then BAR was actually supplied.
To support functions where such empty expansion is allowed, then a new
helper macro `zephyr_check_arguments_required_allow_empty()` has been
implemented, as to be able to distinguish `foo_func()` from
`foo_func(BAR)` when parsing arguments.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
